spec.md

```diff
@@ -151,6 +151,21 @@ Service Broker MAY reject the request with `412 Precondition Failed` and
 provide a message that informs the operator of the API version that is to be
 used instead.
 
+## Extending Resources
+
+Senders of messages defined by this specification MAY include additional
+fields within the JSON objects as extensions. For backwards compatibility
```

Comment:
I really disagree with the application of backward compatibility as a concept here, and suggest that we think the approach.

Here's what I suggest:

- There should be a first-class place, that is 100% explicit on every single message, where extension data goes on payloads that people can reason about as users and authors of the OSB spec
  - You can reason about these specific places and know you're reasoning about the right scope for what OSB is
  - We should make extremely specific guarantees about what the OSB spec says can happen for those specific places
- If you choose to add random extra fields to requests where the extension points are, that's fine, and you can do that, but the first-class place to put extra data for OSB payloads is in the designated spots
- If you implement a client of this API, you only need to implement the designated, explicit places in the API
- If you program to this API, you only need to understand the designated, explicit places for extensions

My concerns with the specific change of saying that brokers can add extensions anywhere:

- It makes the API much harder to reason about as a platform:
  - What data will be present on any particular response from a broker?
  - What expectations will platforms have about what the platform will do with this data?
  - What guarantees does the platform provide to brokers about this random extension data?
- It makes the spec harder to reason about as a broker author, for the same reasons as above
- It complicates platform implementations: now a platform has to deal with the problem of processing unknown fields on **every response from a broker**
- Makes future linting of brokers impossibly difficult: how do you determine if a broker is using these random additions to the spec in a way that is compliant or a good practice?
- I worry about platforms gradually accreting business logic around specific extensions, which provide an uneven experience for broker implementors

On a more general level, I really worry that if we enshrine things that people decided to do without feedback into the WG and going through the normal process, that we are setting a precedent that will constrain our ability to make progress and changes to the spec without breaking someone. We already have confusing areas of the spec around accommodating stateless brokers; I worry that we are potentially opening the door here to even more confusing semantics.

In any event, it's in our best interest to clearly define:

- What the guarantees around extension fields are - for example, if you put extra fields on a broker response, do you ever expect to have them sent back to you?
- What kind of semantics platforms and broker authors can assign to extensions - is the user ever guaranteed to see them? Should platforms ignore them? etc

I really want to foster innovation in this spec and use of extensions. I also don't want to make things that people are doing 'wrong' and out of compliance with the spec. However, I don't think we should say that you can do whatever you want to any request, potentially ascribing any meaning you want to it, and that it's in scope for OSB.

As a practical matter, let me close this with an example. Say that a broker adds some new ID to provision, update, and deprovision requests as a sibling to other top-level fields in the request body. When first-class extension points are adopted into the spec, if this broker does not change at all, they do not suddenly lose compliance with the spec, they are simply not sending OSB extensions. If they want to send that data as OSB extensions, they should change their broker, or otherwise find a way, possibly through a proxy, to **also** send that data in the write places to be considered extensions to OSB.  Nothing will have changed for the broker - I know that Kubernetes does not in any way support additional fields beyond the official ones, and I have no indication that CF does either - they are no more wrong or right in their use of OSB as a spec than before official extentions existed, because they aren't sending official extensions.

Reply:
@pmorie I think you're mixing a lot of topics here. And its way too long to hit all of them, so let me just make a few points:
- regardless of which version of this PR you look at, no one is suggesting that any of these extensions be required for a receiver to understand. The sender has no guarantee that the receiver will understand any of them. In that respect they are all 100% optional and its "sender beware". That's just the nature of extensions.
- extensions, by definition, will not go thru the WG - if they did then they wouldn't be extensions (unless we get into defining "well known extensions", but that's not really the point of this PR).
- tooling on either side of the wire is free to support, or not, the idea of randomly placed extensions. It would then be up to the users of those tools to pester the tool authors to add support if they needed it.
- I don't understand your example but let me add that I don't think the location of the extension has any impact on the semantics behind it - its just a syntax difference.

To me this comes down to a very simple question: what do we want extensions to look like in v3 (or if we were starting from scratch)?

If the answer is: they MAY go anywhere, then that's what we should do now for v2. And I can extract the old version of this PR that just adds a paragraph to say so.

If the answer is: they MUST go into well defined locations in the JSON, then we should not break existing implementations (by allowing them anywhere) but strongly push people towards the v3 model - meaning MAY go anywhere but SHOULD be put into the well defined spots.  The current version of this PR.